pkg/vm/process/message.go

```diff
@@ -0,0 +1,220 @@
+		return "MsgMinValueUnsigned"
+	case MsgMaxValueSigned:
+		return "MsgMaxValueSigned"
+	case MsgMaxValueUnsigned:
+		return "MsgMaxValueUnsigned"
+	}
+	return "unknown message type"
+}
+
+func NewMessageBoard() *MessageBoard {
+	m := &MessageBoard{
+		Messages: make([]*Message, 0, 1024),
+		RwMutex:  &sync.RWMutex{},
+		Mutex:    &sync.Mutex{},
+	}
+	m.Cond = sync.NewCond(m.Mutex)
+	return m
+}
+
+type MessageAddress struct {
+	cnAddr     string
+	pipelineID int32
+	parallelID int32
+}
+
+type Message struct {
+	tag      int32
+	msgType  MsgType
+	sender   *MessageAddress
+	receiver *MessageAddress
+	data     any
+}
+
+type MessageBoard struct {
+	Messages []*Message
+	RwMutex  *sync.RWMutex // for nonblock message
+	Mutex    *sync.Mutex   // for block message
+	Cond     *sync.Cond    //for block message
+}
+
+type MessageReceiver struct {
+	offset  int32
+	tag     int32
+	msgType MsgType
+	addr    *MessageAddress
+	mb      *MessageBoard
+}
+
+func DeepCopyMessageAddr(m *MessageAddress) *MessageAddress {
+	if m == nil {
+		return nil
+	}
+	return &MessageAddress{
+		cnAddr:     m.cnAddr,
+		pipelineID: m.pipelineID,
+		parallelID: m.parallelID,
+	}
+}
+
+func DeepCopyMessage(m *Message) *Message {
+	return &Message{
+		tag:      m.tag,
+		msgType:  m.msgType,
+		sender:   DeepCopyMessageAddr(m.sender),
+		receiver: DeepCopyMessageAddr(m.receiver),
+		data:     m.data,
+	}
+}
+
+func NewMessage(tag int32, msgType MsgType, data any, sender *MessageAddress, receiver *MessageAddress) *Message {
+	return &Message{
+		tag:      tag,
+		msgType:  msgType,
+		sender:   sender,
+		receiver: receiver,
+		data:     data,
+	}
+}
+
+func (proc *Process) SendMessage(m *Message) {
+	mb := proc.MessageBoard
+	mb.RwMutex.Lock()
+	defer mb.RwMutex.Unlock()
+	if m.receiver == nil || m.receiver.cnAddr == CURRENTCN { // message for current CN
+		mb.Messages = append(mb.Messages, m)
+		if messageBlockTable[m.msgType] {
+			// broadcast for block message
+			mb.Cond.Broadcast()
+		}
+	} else {
+		//todo: send message to other CN, need to lookup cnlist
+		panic("unsupported message yet!")
+	}
+}
+
+func (proc *Process) AddrBroadCastToCurrentCN() *MessageAddress {
+	return nil
+}
+
+func (proc *Process) AddrBroadCastToALLCN() *MessageAddress {
+	return &MessageAddress{cnAddr: ALLCN}
+}
+
+func (proc *Process) AddrBroadCastToOneCN(addr string) *MessageAddress {
+	return &MessageAddress{cnAddr: addr}
+}
+
+func (proc *Process) AddrBroadCastToOneParallel(addr string, pipelineid int32, parallelid int32) *MessageAddress {
+	return &MessageAddress{
+		cnAddr:     addr,
+		pipelineID: pipelineid,
+		parallelID: parallelid,
+	}
+}
+
+func (proc *Process) NewMessageReceiver(tag int32, msgType MsgType, addr *MessageAddress) *MessageReceiver {
+	return &MessageReceiver{
+		tag:     tag,
+		msgType: msgType,
+		addr:    addr,
+		mb:      proc.MessageBoard,
+	}
+}
+
+func (mr *MessageReceiver) receiverMessageNonBlock(result []*Message) {
+	mr.mb.RwMutex.RLock()
```

Comment:
First, tag is not useful.
Second, why do you care about message type.    Just recv and let receiver handle message type.    In a sense, I think message type is also useless.    Message is data, agreed by sender and receiver.